book/src/developers/designs/identity-verification.md

```diff
@@ -0,0 +1,152 @@
+## Draft proposal for and Identity verification system with an out of band challenge
+
+Implements #337
+
+### Goal:
+
+Providing a way to allow 2 subject to authenticate each other using an out of band challenge.
+
+### Assumptions:
+
+Both subjects will have to be connected to a Kanidm server (not necessarily the same or under the same domain), and the servers will have to mutually trust each other; the trust bond must be established beforehand.
+
+### High level overview:
+
+The idea is to associate a [ECDH](https://docs.rs/openssl/latest/openssl/pkey_ctx/struct.PkeyCtxRef.html#method.derive_set_peer) public/secret key pair with each user. If two users want to authenticate each other, their private key and the other subject's public keys will be used to encrypt an HMAC message that uses both of the users UUID. Kanidm currently doesn't support ECDH keys for users, but once they will be implemented we will be able to use them for other purposes as well.
```

Comment:
I don't know why you keep saying "hmac" or "encryption" here. Neither of these are involved in this process. 

I think you should try implementing this but in isolation to "see" the process.

The steps will be.

* generate a private and public ec key for both parties.
* exchange of the public keys (that is taken care of in kanidm for you by the underlying replication/database systems)
* each side does the derivation:
** given entity A and it's private key, use B's public key with ecdh to get shared secret X
** given entity B and it's private key, use A's public key with ecdh to get shared secret X

Then each side does the same set of derivations to get TWO TOTP secrets.

both sides will perform:

* concatenate the bytes of secret X with entity A uuid as bytes. This is TOTP secret A.
* concatenate the bytes of secret X with entity B uuid as bytes. This is TOTP secret B.

Both sides now have the two secrets. 

Entity A is displayed the output of TOTP(totp secret A).
Entity B is displayed the output of TOTP(totp secret B).

Entity B reads out their displayed TOTP B to entity A who types that value in to validate it. 
Entity A reads out their displayed TOTP A to entity B who types that value in to validate it.


This satisfies all our properties:

* The secret X is unique to the combination of A and B validating each other.
* By concatenating X with A/B uuids, we have two seperate secrets for each side to validate the other. 
* TOTP gives us the rolling / time changing of secrets with a grace window to allow some forgiveness in the exchange. We'll likely have to make the TOTP step 10 minutes in this scheme.
* There is no online exchange or communication
* There is no encryption or hmac going on.

So can you update the document to reflect these steps?

Reply:
I think I should remove the overview section altogether as it fails explains the steps I detail afterwards. By encryption I meant just using the keys for the secret derivation the way you described, no online exchange involved, so I think we're on the same boat here. As in regard for hashing, I thought it could be an alternative to just concatenate the UUID as you suggested. This way, we would hash the shared secret with the user's UUID and use that as a TOTP, hopefully providing better security as it wouldn't be possible to retrive the UUID from the totp.

Reply:
Update: I realized I never mentioned what the TOTP was going to be, now I hope it's much more clear :)

Reply:
I was creating the proposal on the Google website: what project size should I pick? They offer just medium or large.

Reply:
pick large :)

Reply:
TOTP internally already does the HMAC for you, that's why concatenation of the bytes is sufficient.

Reply:
I see now, so there's no need to mention hmac in the design at all

Reply:
correct!